Add circuit relay features to the networking.

shamil-gadelshin opened the pull request:
This PR introduces libp2p circuit relay features to the `subspace-networking` crate, it implements the networking prototype for [single port for farmers improvement](https://github.com/subspace/subspace/issues/612).

New features support relaying via `MemoryTransport` however it's not a requirement. 

The current prototype assumes three relaying roles: relay server node, relay client with no external address (circuit relay reserver), and relay client-aware node (circuit relay initiator). Complex scenarios with multiple listening addresses for different nodes with or without circuits weren't tested.

#### Changes

- new relay (server) and relay client configurations
- new relay limit settings helper struct which is crucial to circuit relay creating rates
- relay client and memory transport support
- new relay and relay client behaviour fields and corresponding events

#### Testing

- `relayed-requests` - tests relayed request-response protocol
- `relayed-networking` - tests relayed gossipsub protocol
- `relayed-get-complex-pieces` - tests complex kademlia protocol setup

#### Comments

- the changes are backward compatible with a disabled relay configuration by default
- having a configured Node in the `RelayConfiguration::Server` mode allows to create a relay client configuration using `node.configure_relay_client()` method, which is helpful to our initial relay circuits use case
- heavy circuit relay configuration required changing the default relay limit settings (see `relayed-get-pieces-complex` example)
- `RelayLimitSettings` struct is unnecessary to the implementation and was introduced only to counter `relay::Config` struct limitations (not Clone, Send, Sync). We could probably create a libp2p refactoring PR.
- `DEFAULT_RELAY_SERVER_ADDRESS` static field was introduced to simplify the usage of the new features, however, it's not required to the implementation and could be removed.

shamil-gadelshin committed:
networking: Introduce libp2p relay support.

- add relay (server) and relay client configurations
- add relay limit settings helper struct
- add relay client and memory transport support
- add relay and relay client behaviours and corresponding events

shamil-gadelshin committed:
networking: Add relay events tracing.

shamil-gadelshin committed:
networking: Export new relay types.

shamil-gadelshin committed:
networking: Add new `relayed-requests` example

shamil-gadelshin committed:
networking: Add new `relayed-networking` example

shamil-gadelshin committed:
networking: Add new `relayed-get-pieces-complex` example

shamil-gadelshin committed:
networking: Fix relay configuration comments.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -30,6 +38,153 @@ use tracing::info;
+/// Provides the default relay server multi-address. It's memory based and can be used in both
+/// development and in-memory relay scenarios.
+pub static DEFAULT_RELAY_SERVER_ADDRESS: Lazy<Multiaddr> =
+    Lazy::new(|| Multiaddr::empty().with(Protocol::Memory(1_000_000_000)));
```
I don't really see the benefit of it if we clone it later anyway. Also Shouldn't the port be `0` to make it pick random? Otherwise two instances in the same process will collide when trying to use the same port.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -30,6 +38,153 @@ use tracing::info;
+        let circuit_src_rate_limiters = vec![
+            rate_limiter::new_per_peer(self.circuit_src_rate_limit_per_peer),
+            rate_limiter::new_per_ip(self.circuit_src_rate_limit_per_ip),
+        ];
```
I don't think we need this. The only restriction must be to only allow local peers (ideally just memory transport or identify them by known PeerId), the rest should have access denied, so we probably don't need rate limit configs like above.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -30,6 +38,153 @@ use tracing::info;
+    /// It will enable creating a relay circuit.
+    /// Expected format for Multiaddr: /ip4/127.0.0.1/tcp/50000/p2p/<server_peer_id>/p2p-circuit
+    /// This option will toggle the relay-client behaviour.
+    ClientInitiator,
```
I find these explanations confusing.

First of all, server behind NAT is strange, usually it is clients that are behind NAT, not servers.

Second is `/ip4/127.0.0.1/tcp/50000/p2p/<server_peer_id>/p2p-circuit`, usually `p2p-circuit` isn't the end, there is `/p2p/QmAlice` after it.

Third, there is expected multiaddr format description for `ClientInitiator`, but no such parameter in it (likely copy-paste).

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -253,3 +421,60 @@ pub async fn create(
     Ok((node, node_runner))
 }
+
+// Builds the transport stack that LibP2P will communicate over along with an optional relay client.
+async fn build_transport(
+    keypair: identity::Keypair,
+    timeout: Duration,
+    yamux_config: YamuxConfig,
+    relay_transport: Option<ClientTransport>,
+) -> Result<Boxed<(PeerId, StreamMuxerBox)>, CreationError> {
```
Nit: it would have been preferable to extract this function in a separate commit. Then it is much easier to follow what has actually changes.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -253,3 +421,60 @@ pub async fn create(
+            timeout,
+            yamux_config,
+        ))
+    }
```
A bit cleaner:
```suggestion
    Ok(upgrade_transport(
        if let Some(relay_transport) = relay_transport {
            OrTransport::new(relay_transport, transport).boxed()
        } else {
            transport.boxed()
        },
        &keypair,
        timeout,
        yamux_config,
    ))
```

Or leave as is and make the whole `transport` argument generic in `upgrade_transport`, so you don't have to box it here at all (returned upgraded transport is boxed anyway and will result in the same type anyway, but one more pointer indirection will be eliminated).

shamil-gadelshin replied:
The proposed code is not compilable, I introduced complex generic arguments instead.

nazar-pc commented on crates/subspace-networking/src/node.rs:
```diff
@@ -114,22 +116,48 @@ pub enum SendPiecesByRangeRequestError {
+                    .with(Protocol::P2pCircuit),
+            ))
+        } else {
+            Err(RelayConfigurationError::ExpectedServerConfiguration)
```
Yeah, not particularly nice, I think we'll end up with a separate behavior in the end to avoid this.

nazar-pc commented on crates/subspace-networking/examples/relayed-requests.rs:
```diff
@@ -0,0 +1,102 @@
+
+    // NODE 1 - Relay
+    let node_1_addr: Multiaddr = "/ip4/127.0.0.1/tcp/50000".parse().unwrap();
+    let custom_relay_address: Multiaddr = "/memory/10".parse().unwrap();
```
Random should be preferred if possible:
```
     let custom_relay_address: Multiaddr = "/memory/0".parse().unwrap(); 
```

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -144,7 +144,6 @@ pub enum RelayConfiguration {
 
     /// Relay-client configuration in requesting mode (a client for a server behind NAT).
     /// It will enable creating a relay circuit.
-    /// Expected format for Multiaddr: /ip4/127.0.0.1/tcp/50000/p2p/<server_peer_id>/p2p-circuit
```
This last commit is exactly the kind of change I'd squash with corresponding earlier commit before pushing (or force-pushed if already submitted), it doesn't really deserve its own commit.

shamil-gadelshin replied:
I agree. However, I found it in the last moments on the PR creation (self-review) after all the previous squashes. I'll force-push the branch next time.

nazar-pc reviewed (commented):
Have some questions and comments, but overall looks good and makes sense to me.

I'm not sure we need to clone all examples though, one simple example that showcases relay specifically would be sufficient, else we'll have to maintain multiple copies of basically the same code.

i1i1 commented on crates/subspace-networking/src/behavior.rs:
```diff
@@ -120,3 +137,15 @@ impl From<RequestResponseEvent> for Event {
+    fn from(event: RelayClientEvent) -> Self {
+        Event::RelayClient(event)
+    }
+}
```
Maybe we can add [`derive(derive_more::From)`](https://jeltef.github.io/derive_more/derive_more/from.html) for `Event`. There are already 7 variants and 7 from impls

shamil-gadelshin committed:
networking: Fix review comments

- delete redundant test (relayed-networking)
- remove default server address for relays
- improve comments
- remove redundant rate limits
- add derive_from macro usage
- add complex generics to upgrade_transport()

shamil-gadelshin committed:
Merge branch 'main' into relay-circuits

shamil-gadelshin committed:
networking: Fix merge issues.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -458,14 +408,25 @@ async fn build_transport(
+            ListenerUpgrade = ListenerUpgrade,
+            Dial = Dial,
+        > + Send
+        + 'static,
```
Let's prefer generics over `impl`, it is easier to read function signature that way

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -73,55 +62,23 @@ impl Default for RelayLimitSettings {
 impl RelayLimitSettings {
     /// Converts `RelayLimitSettings` to libp2p `relay::Config`.
     pub fn to_relay_config(self) -> RelayConfig {
-        let reservation_rate_limiters = vec![
```
We removed rate limiters, but I don't think we have any restriction yet on who can use this circuit relay. The intention is to limit it just to the local peers within the same process. Probably worth having a TODO so we don't forget about it.

nazar-pc reviewed (dismissed):
Overall looks fine as the first version

shamil-gadelshin committed:
Merge branch 'main' into relay-circuits

shamil-gadelshin committed:
networking: Fix review comments

- add TODO for relay rate limits
- change impl parameter to generic for upgrade_transport()
